### ISSUE TITLE
Questions about supported cloud storage.

### ISSUE THREAD

**Issue by manoaman** [google/neuroglancer#319]
Hi,

Where can I find a list of supported cloud storage services I can find except AWS and GoogleCloud Storage?

https://github.com/google/neuroglancer/blob/master/src/neuroglancer/datasource/file_protocols.md

Would it be possible to use Microsoft Box, Google Drive, or even file servers (Apache, nginx) on bare-metals or vm servers for production use?  If yes, are there suggested authentication methods compatible with Neuroglancer?

Thank you,
-m

**Comment by jbms** [comment]
You can use any web server as long as it will serve appropriate CORS headers, so many services could be supported.

Google drive does not currently work, though it could potentially be supported.

In general unauthenticated will be much easier to support than authenticated, due to additional security issues related to neuroglancer being a web application.  If you want to try to support authentication for a particular service we can discuss that, though. 

In many cases instead of user authentication you can instead rely on a "capability url", where the url itself contains a long random string.  For example, you can use a bucket name containing a long random string but make the bucket publicly readable.  Then only someone who knows the bucket name can access it.

**Comment by manoaman** [comment]
Hi @jbms ,

When considering AWS S3 storage, and estimating the cost, is GET request the only communication from the Neuroglancer that should be considered?  Also, if chunk sizes are designed smaller, more requests will be made to the S3 storage?  Is there a ballpark number of requests I can estimate from the file size?

thanks,
-m

**Comment by jbms** [comment]
Yes, you only need to worry about the GET requests.  The only other request Neuroglancer makes is bucket list requests, but that is only done to complete paths when typing in a data source interactively, so those should be negligible.

Smaller chunk size will lead to more requests, but a larger chunk size may lead to excess data transfer.  I would recommend a chunk size of about 64x64x64 if the resolution is isotropic.

To estimate the number of requests you can just look at the Network tab within Chrome Dev Tools after browsing around.

**Comment by manoaman** [comment]
Chrome browser has a feature to "Disable cache".  Does Neuroglancer have an ability to cache to reduce the number of requests to the backend storage?  Or could Neuroglancer be caching on its own whenever new requests are made from the client?  I wanted to know if literally the number of requests within Chrome Dev Tools is counted towards "requests" made to the S3 storage.

**Comment by jbms** [comment]
Neuroglancer doesn't have its own cache, but the browser may cache requests.  Note though that currently due to https://bugs.chromium.org/p/chromium/issues/detail?id=969828, when using Chrome with the neuroglancer precomputed sharded format, requests are not cached.  However, other data formats, and Firefox, are not affected.

**Comment by jbms** [comment]
As long as your chunks are ~100KB or greater, I think you will find that egress charges are greater than the per-operation cost, so increasing your chunk size further won't help in that regard.  You could consider a different provider that has lower egress and/or operation charges.

**Comment by manoaman** [comment]
Hi @jbms 

Suppose a single 3D volume (1 brain) is constructed with 15 million files, is it safe to say that 15 million requests at minimal is required to view the entire brain on the Neuroglancer?  In reality, I think the user will iteratively navigate x-y-z planes along with zoom in/out so I would think much higher number.  Am I assuming it wrong?

Thanks,
-m

**Comment by jbms** [comment]
Normally for a large volume you would want to also have downsampled levels in addition to the full resolution, and I would expect that the a user would only view small portions at full resolution.  The best way to get an estimate is to browse around using Neuroglancer and check the total amount of data transferred, using the Dev Tools network tab.

If you had 15 million files of 64x64x64 uint8 uncompressed chunks, that would be `15e6 * 64^3 = 3.93 TB` of data in total.  To download that much data would take a very long time.

**Comment by manoaman** [comment]
It sounds like I need more effort to downsize the data in general.  Thank you for your advise @jbms 

By the way, do you happen to know if Google offers a data sponsorship program?  It came to my attention aws offers this program and I was wondering if there is something equivalent.

https://aws.amazon.com/opendata/open-data-sponsorship-program/

Thanks,
-m

**Comment by jbms** [comment]
To be clear, 3.93TB is not necessarily too large for the complete dataset, but you will definitely want to have downsampled levels in addition to the full resolution in order to be able to efficiently zoom out in Neuroglancer.  I was just pointing out, for the purpose of calculating egress costs, that no single user is going to download that much data in a single browsing session.  More likely a user might download 1GB or so in a browsing session.

As far as the AWS open data sponsorship program, I am not aware of a similar program at Google, though it is possible there is one.

**Comment by awsmrud** [comment]
Can Neuroglancer be used to fetch data from Amazon S3 private buckets?

**Comment by jbms** [comment]
Not yet, a solution for authentication needs to be developed.
